### PR TITLE
feat(performance): sampling support for funnels

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Problem

#12908 

## Changes

Adds sampling support for funnels + some refactoring.

Moving fast here but overall there are a few questions about sampling that we need to answer (see https://github.com/PostHog/posthog/issues/12908#issuecomment-1433573133)

## How did you test this code?

Manually

### CODE REVIEW

File: posthog/queries/funnels/funnel_trends.py [hunk 1]
```diff
@@ -162,11 +167,12 @@ def _summarize_data(self, results):
...
 
         summary = []
+
         for period_row in results:
             serialized_result = {
                 "timestamp": period_row[0],
-                "reached_from_step_count": period_row[1],
-                "reached_to_step_count": period_row[2],
+                "reached_from_step_count": correct_result_for_sampling(period_row[1], self._filter.sampling_factor),
+                "reached_to_step_count": correct_result_for_sampling(period_row[2], self._filter.sampling_factor),
```
Comment (EDsCODE) [inline posthog/queries/funnels/funnel_trends.py:175]: Might make sense to include the correction in the query itself. Query classes can piggy back off `self._filter.sampling_factor` and add the adjustment if necessary. Then none of these count format functions need to consider this

Comment (EDsCODE) [reply posthog/queries/funnels/funnel_trends.py:175]: Just saw the issue and consideration for this that happened already!

File: frontend/src/scenes/insights/EditorFilters/SamplingFilter.tsx [hunk 2]
```diff
@@ -5,29 +5,54 @@ import { insightLogic } from 'scenes/insights/insightLogic'
...
-    const { setFilters } = useActions(insightLogic)
-    const { filters } = useValues(insightLogic)
     const { featureFlags } = useValues(featureFlagLogic)
 
     // Sampling is currently behind a feature flag and only available on lifecycle queries
-    if (!featureFlags[FEATURE_FLAGS.SAMPLING] || editorFilters.insight !== InsightType.LIFECYCLE) {
-        return <></>
+    const insightSupportsSampling =
+        (featureFlags[FEATURE_FLAGS.SAMPLING] && editorFilters.insight === InsightType.LIFECYCLE) ||
+        editorFilters.insight === InsightType.FUNNELS
```
Comment (pauldambra) [inline frontend/src/scenes/insights/EditorFilters/SamplingFilter.tsx:26]:
```suggestion
        featureFlags[FEATURE_FLAGS.SAMPLING] && (editorFilters.insight === InsightType.LIFECYCLE ||
        editorFilters.insight === InsightType.FUNNELS)
```

Comment (pauldambra) [reply frontend/src/scenes/insights/EditorFilters/SamplingFilter.tsx:26]: Brackets in wrong place?

As it was sampling was always visible for funnels even with flag off

**Review by pauldambra — APPROVED** [review]
One possible change but otherwise more than good enough for some testing and checking 👍

File: frontend/src/scenes/insights/EditorFilters/SamplingFilter.tsx (hunk 2 above)
Comment (yakkomajuri) [reply frontend/src/scenes/insights/EditorFilters/SamplingFilter.tsx:26]: ah yup I forgot the parentheses so the linter added these - but indeed!

File: posthog/queries/funnels/funnel_trends.py (hunk 1 above)
Comment (EDsCODE) [reply posthog/queries/funnels/funnel_trends.py:175]: Fair, I suppose if there's no real performance gains for doing it query-side, python side makes sense

Comment (yakkomajuri) [reply posthog/queries/funnels/funnel_trends.py:175]: Coming back to this discussion to say that while I feel strongly about doing more on the Python side, I do think it's always a case by case thing so will be considering how to best approach sampling going forward.

I thought it'd make things messy to do math and handle zeros and the like on the query side, further polluting already complicated queries. But I might give a query a shot and see how it turns out. It definitely ensures we don't miss a spot later down where we need to account for sampling

**Pull request merged**